## Setup: Import Libraries and Load Data

This is my first Python test:
print("Hello World! -- explore this")

In [2]:
!pip install pandas


[notice] A new release of pip is available: 25.3 -> 26.0
[notice] To update, run: pip install --upgrade pip


In [3]:
# Import the essential libraries
import pandas as pd
import numpy as np

# Display settings for better output
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)

print("Libraries imported successfully!")
print(f"pandas version: {pd.__version__}")
print(f"numpy version: {np.__version__}")

Libraries imported successfully!
pandas version: 2.3.3
numpy version: 2.4.1


In [4]:
df = pd.read_csv("diabetes.csv")
df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [5]:
# Load the Diabetes dataset
df = pd.read_csv('diabetes.csv')

print("Dataset loaded successfully!")

Dataset loaded successfully!


---

# The 10-Point Inspection

Like a pilot's pre-flight checklist, we go through each point methodically before taking off with our analysis.

| Point | What We Check | Command |
|-------|---------------|----------|
| 1 | Shape | `df.shape` |
| 2 | Columns | `df.columns` |
| 3 | Data Types | `df.dtypes` |
| 4 | First Look | `df.head()` |
| 5 | Last Look | `df.tail()` |
| 6 | Memory | `df.memory_usage()` |
| 7 | Missing | `df.isnull().sum()` |
| 8 | Duplicates | `df.duplicated()` |
| 9 | Statistics | `df.describe()` |
| 10 | Unique | `df.nunique()` |

---

## Point 1: Shape

**Question:** How big is our dataset?

The shape tells us:
- **Rows** = Number of observations (students in our case)
- **Columns** = Number of features/variables

In [6]:
# Point 1: Shape
print("Dataset Shape:")
print(df.shape)

print(f"\n📊 We have {df.shape[0]:,} observations and {df.shape[1]} features")
print(f"📊 Total data points: {df.shape[0] * df.shape[1]:,}")

Dataset Shape:
(768, 9)

📊 We have 768 observations and 9 features
📊 Total data points: 6,912


---

## Point 2: Column Names

**Question:** What features do we have?

Understanding column names helps us:
- Know what data is available
- Identify potential issues (spaces, special characters)
- Start thinking about feature relationships

In [8]:
# Point 2: Column Names
print("Column Names:")
print(df.columns.tolist())

print(f"\n📋 Total columns: {len(df.columns)}")

Column Names:
['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome']

📋 Total columns: 9


---

## Point 3: Data Types

**Question:** How is each column stored?

⚠️ **KEY INSIGHT:** Just because pandas says `int64` doesn't mean it should be treated as numeric!

In [9]:
# Point 3: Data Types
print("Data Types:")
print(df.dtypes)

print("\n" + "="*50)
print("Data Type Summary:")
print(df.dtypes.value_counts())

Data Types:
Pregnancies                   int64
Glucose                       int64
BloodPressure                 int64
SkinThickness                 int64
Insulin                       int64
BMI                         float64
DiabetesPedigreeFunction    float64
Age                           int64
Outcome                       int64
dtype: object

Data Type Summary:
int64      7
float64    2
Name: count, dtype: int64


---

## Point 4: First Look (Head)

**Question:** What do the first few rows look like?

This helps us:
- See actual data values
- Spot obvious data quality issues
- Understand the data format

In [10]:
# Point 4: First Look
print("First 5 Rows:")
df.head()

First 5 Rows:


Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

---

## Point 5: Last Look (Tail)

**Question:** What do the last few rows look like?

Why check the tail?
- Data might be sorted or ordered
- End of file might have different patterns
- Can reveal data entry issues

In [11]:
# Point 5: Last Look
print("Last 5 Rows:")
df.tail()

Last 5 Rows:


Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0

---

## Point 6: Memory Usage

**Question:** How much memory does our dataset use?

Important for:
- Knowing if data fits in RAM
- Planning for larger datasets
- Optimizing data types if needed

In [12]:
# Point 6: Memory Usage
print("Memory Usage by Column:")
print(df.memory_usage(deep=True))

# Total memory in MB
total_memory_mb = df.memory_usage(deep=True).sum() / 1e6
print(f"\n💾 Total Memory Usage: {total_memory_mb:.2f} MB")

Memory Usage by Column:
Index                        132
Pregnancies                 6144
Glucose                     6144
BloodPressure               6144
SkinThickness               6144
Insulin                     6144
BMI                         6144
DiabetesPedigreeFunction    6144
Age                         6144
Outcome                     6144
dtype: int64

💾 Total Memory Usage: 0.06 MB


---

## Point 7: Missing Values

**Question:** Do we have any missing data?

Missing data is crucial because:
- Many algorithms can't handle NaN values
- Missing patterns can be informative
- We may need imputation strategies

In [13]:
# Point 7: Missing Values
print("Missing Values by Column:")
missing = df.isnull().sum()
print(missing)

print("\n" + "="*50)
total_missing = missing.sum()
print(f"❓ Total Missing Values: {total_missing}")

if total_missing == 0:
    print("✅ Great! No missing values - complete dataset!")
else:
    print(f"⚠️ {total_missing} missing values need attention")

Missing Values by Column:
Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

❓ Total Missing Values: 0
✅ Great! No missing values - complete dataset!


---

## Point 8: Duplicate Rows

**Question:** Do we have any duplicate records?

Duplicates can:
- Inflate sample size artificially
- Bias statistical analyses
- Indicate data collection issues

In [14]:
# Point 8: Duplicates
duplicate_count = df.duplicated().sum()
duplicate_pct = (duplicate_count / len(df)) * 100

print(f"📑 Duplicate Rows: {duplicate_count:,}")
print(f"📑 Percentage: {duplicate_pct:.2f}%")

if duplicate_count > 0:
    print(f"\n⚠️ Warning: {duplicate_pct:.2f}% of rows are duplicates!")
    print("   This needs investigation in Week 4 (Data Cleaning)")

📑 Duplicate Rows: 0
📑 Percentage: 0.00%


In [15]:
# Let's look at some duplicate examples
print("Example of duplicate rows:")
duplicates = df[df.duplicated(keep=False)]
duplicates.head(10)

Example of duplicate rows:


Empty DataFrame
Columns: [Pregnancies, Glucose, BloodPressure, SkinThickness, Insulin, BMI, DiabetesPedigreeFunction, Age, Outcome]
Index: []

---

## Point 9: Descriptive Statistics

**Question:** What are the statistical properties of our numeric columns?

This tells us:
- Central tendency (mean, median)
- Spread (std, min, max)
- Distribution hints (quartiles)

In [16]:
# Point 9: Descriptive Statistics
print("Descriptive Statistics:")
df.describe()

Descriptive Statistics:


Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

In [17]:
# Transpose for easier reading
print("Descriptive Statistics (Transposed for readability):")
df.describe().T

Descriptive Statistics (Transposed for readability):


count        mean         std     min       25%  \
Pregnancies               768.0    3.845052    3.369578   0.000   1.00000   
Glucose                   768.0  120.894531   31.972618   0.000  99.00000   
BloodPressure             768.0   69.105469   19.355807   0.000  62.00000   
SkinThickness             768.0   20.536458   15.952218   0.000   0.00000   
Insulin                   768.0   79.799479  115.244002   0.000   0.00000   
BMI                       768.0   31.992578    7.884160   0.000  27.30000   
DiabetesPedigreeFunction  768.0    0.471876    0.331329   0.078   0.24375   
Age                       768.0   33.240885   11.760232  21.000  24.00000   
Outcome                   768.0    0.348958    0.476951   0.000   0.00000   

                               50%        75%     max  
Pregnancies                 3.0000    6.00000   17.00  
Glucose                   117.0000  140.25000  199.00  
BloodPressure              72.0000   80.00000  122.00  
SkinThickness              23.0000   32.00000   99.00  
Insulin                    30.5000  127.25000  846.00  
BMI                        32.0000   36.60000   67.10  
DiabetesPedigreeFunction    0.3725    0.62625    2.42  
Age                        29.0000   41.00000   81.00  
Outcome                     0.0000    1.00000    1.00

---

## Point 10: Unique Values

**Question:** How many unique values does each column have?

This reveals:
- Binary features (2 unique values)
- Categorical features (few unique values)
- Continuous features (many unique values)

In [18]:
# Point 10: Unique Values
print("Unique Values per Column:")
unique_counts = df.nunique().sort_values()
print(unique_counts)

Unique Values per Column:
Outcome                       2
Pregnancies                  17
BloodPressure                47
SkinThickness                51
Age                          52
Glucose                     136
Insulin                     186
BMI                         248
DiabetesPedigreeFunction    517
dtype: int64


In [19]:
# Categorize columns by unique count
print("\n" + "="*50)
print("Feature Classification by Unique Values:")
print("="*50)

binary = unique_counts[unique_counts == 2].index.tolist()
low_cardinality = unique_counts[(unique_counts > 2) & (unique_counts <= 5)].index.tolist()
high_cardinality = unique_counts[unique_counts > 5].index.tolist()

print(f"\n🟢 BINARY (2 values): {binary}")
print(f"\n🔵 LOW CARDINALITY (3-5 values): {low_cardinality}")
print(f"\n🟣 HIGH CARDINALITY (>5 values): {high_cardinality}")


Feature Classification by Unique Values:

🟢 BINARY (2 values): ['Outcome']

🔵 LOW CARDINALITY (3-5 values): []

🟣 HIGH CARDINALITY (>5 values): ['Pregnancies', 'BloodPressure', 'SkinThickness', 'Age', 'Glucose', 'Insulin', 'BMI', 'DiabetesPedigreeFunction']


---

# Feature Classification Summary

Based on our 10-Point Inspection, we can classify our features:

In [20]:
# Feature Classification
feature_classification = {
    'Numeric (Discrete Counts)': [
        'Pregnancies (number of times pregnant)',
        'Age (years)'
    ],
    'Numeric (Continuous Measurements)': [
        'Glucose (plasma glucose concentration)',
        'BloodPressure (mm Hg)',
        'SkinThickness (triceps skin fold thickness, mm)',
        'Insulin (2-Hour serum insulin, mu U/ml)',
        'BMI (body mass index)',
        'DiabetesPedigreeFunction (genetic risk score)'
    ],
    'Target Variable': [
        'Outcome (0 = No Diabetes, 1 = Diabetes) — BINARY CLASSIFICATION'
    ]
}

for category, features in feature_classification.items():
    print(f"\n{'='*50}")
    print(f"📌 {category}")
    print('='*50)
    for feature in features:
        print(f"   • {feature}")


📌 Numeric (Discrete Counts)
   • Pregnancies (number of times pregnant)
   • Age (years)

📌 Numeric (Continuous Measurements)
   • Glucose (plasma glucose concentration)
   • BloodPressure (mm Hg)
   • SkinThickness (triceps skin fold thickness, mm)
   • Insulin (2-Hour serum insulin, mu U/ml)
   • BMI (body mass index)
   • DiabetesPedigreeFunction (genetic risk score)

📌 Target Variable
   • Outcome (0 = No Diabetes, 1 = Diabetes) — BINARY CLASSIFICATION


---

# Sanity Checks

**Question:** Do our values make real-world sense?

In [21]:
print("🔍 SANITY CHECKS")
print("="*50)

checks = [
    ('Pregnancies', 'Pregnancies', 0, 20, 'Reasonable lifetime pregnancy count'),
    ('Glucose', 'Glucose', 50, 250, 'Physiologically plausible plasma glucose range'),
    ('BloodPressure', 'BloodPressure', 30, 150, 'Possible diastolic blood pressure range (mm Hg)'),
    ('SkinThickness', 'SkinThickness', 5, 60, 'Plausible triceps skinfold thickness (mm)'),
    ('Insulin', 'Insulin', 0, 900, 'Possible 2-hour serum insulin range'),
    ('BMI', 'BMI', 10, 70, 'Humanly possible BMI range'),
    ('DiabetesPedigreeFunction', 'DiabetesPedigreeFunction', 0.0, 3.0, 'Typical range for genetic risk score'),
    ('Age', 'Age', 18, 90, 'Adult age range')
]

all_passed = True
for name, col, expected_min, expected_max, note in checks:
    actual_min = df[col].min()
    actual_max = df[col].max()
    
    passed = (actual_min >= expected_min - 5) and (actual_max <= expected_max + 5)
    status = "✅" if passed else "❌"
    
    print(f"\n{status} {name}:")
    print(f"   Range: {actual_min} to {actual_max}")
    print(f"   Note: {note}")
    
    if not passed:
        all_passed = False

print("\n" + "="*50)
if all_passed:
    print("✅ ALL SANITY CHECKS PASSED!")
    print("   Data is internally consistent with real-world expectations")
else:
    print("⚠️ Some checks failed — investigate zeros or extreme outliers!")

🔍 SANITY CHECKS

✅ Pregnancies:
   Range: 0 to 17
   Note: Reasonable lifetime pregnancy count

❌ Glucose:
   Range: 0 to 199
   Note: Physiologically plausible plasma glucose range

❌ BloodPressure:
   Range: 0 to 122
   Note: Possible diastolic blood pressure range (mm Hg)

❌ SkinThickness:
   Range: 0 to 99
   Note: Plausible triceps skinfold thickness (mm)

✅ Insulin:
   Range: 0 to 846
   Note: Possible 2-hour serum insulin range

❌ BMI:
   Range: 0.0 to 67.1
   Note: Humanly possible BMI range

✅ DiabetesPedigreeFunction:
   Range: 0.078 to 2.42
   Note: Typical range for genetic risk score

✅ Age:
   Range: 21 to 81
   Note: Adult age range

⚠️ Some checks failed — investigate zeros or extreme outliers!


---

# Reusable 10-Point Inspection Function

Here's a function you can reuse for any dataset:

In [22]:
# Bonus: df.info() provides a quick summary
print("DataFrame Info:")
df.info()

DataFrame Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


In [23]:
def ten_point_inspection(df, name="Dataset"):
    """
    Perform a comprehensive 10-point inspection on a DataFrame.
    
    Parameters:
    -----------
    df : pandas DataFrame
        The dataset to inspect
    name : str
        Name of the dataset for display purposes
    """
    print("="*60)
    print(f"📊 10-POINT INSPECTION: {name}")
    print("="*60)
    
    # 1. Shape
    print(f"\n1️⃣  SHAPE: {df.shape[0]:,} rows × {df.shape[1]} columns")
    
    # 2. Columns
    print(f"\n2️⃣  COLUMNS: {list(df.columns)}")
    
    # 3. Data Types
    print(f"\n3️⃣  DATA TYPES:")
    print(df.dtypes.value_counts().to_string())
    
    # 4 & 5. First and Last rows (just noting)
    print(f"\n4️⃣  FIRST ROW: {dict(df.iloc[0])}")
    print(f"\n5️⃣  LAST ROW: {dict(df.iloc[-1])}")
    
    # 6. Memory
    memory_mb = df.memory_usage(deep=True).sum() / 1e6
    print(f"\n6️⃣  MEMORY: {memory_mb:.2f} MB")
    
    # 7. Missing Values
    missing = df.isnull().sum().sum()
    print(f"\n7️⃣  MISSING VALUES: {missing:,}")
    
    # 8. Duplicates
    dupes = df.duplicated().sum()
    dupe_pct = (dupes / len(df)) * 100
    print(f"\n8️⃣  DUPLICATES: {dupes:,} ({dupe_pct:.2f}%)")
    
    # 9. Key Statistics
    print(f"\n9️⃣  KEY STATISTICS:")
    numeric_cols = df.select_dtypes(include=[np.number]).columns[:5]  # First 5 numeric
    for col in numeric_cols:
        print(f"    {col}: min={df[col].min()}, max={df[col].max()}, mean={df[col].mean():.2f}")
    
    # 10. Unique Values
    print(f"\n🔟 UNIQUE VALUE RANGES:")
    unique = df.nunique()
    print(f"    Binary (2): {list(unique[unique == 2].index)}")
    print(f"    Low (3-5): {list(unique[(unique > 2) & (unique <= 5)].index)}")
    print(f"    High (>5): {list(unique[unique > 5].index)}")
    
    print("\n" + "="*60)
    print("✅ 10-Point Inspection Complete!")
    print("="*60)

In [24]:
# Run the inspection function
ten_point_inspection(df, "Pima Indian Diabetes")

📊 10-POINT INSPECTION: Pima Indian Diabetes

1️⃣  SHAPE: 768 rows × 9 columns

2️⃣  COLUMNS: ['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome']

3️⃣  DATA TYPES:
int64      7
float64    2

4️⃣  FIRST ROW: {'Pregnancies': np.float64(6.0), 'Glucose': np.float64(148.0), 'BloodPressure': np.float64(72.0), 'SkinThickness': np.float64(35.0), 'Insulin': np.float64(0.0), 'BMI': np.float64(33.6), 'DiabetesPedigreeFunction': np.float64(0.627), 'Age': np.float64(50.0), 'Outcome': np.float64(1.0)}

5️⃣  LAST ROW: {'Pregnancies': np.float64(1.0), 'Glucose': np.float64(93.0), 'BloodPressure': np.float64(70.0), 'SkinThickness': np.float64(31.0), 'Insulin': np.float64(0.0), 'BMI': np.float64(30.4), 'DiabetesPedigreeFunction': np.float64(0.315), 'Age': np.float64(23.0), 'Outcome': np.float64(0.0)}

6️⃣  MEMORY: 0.06 MB

7️⃣  MISSING VALUES: 0

8️⃣  DUPLICATES: 0 (0.00%)

9️⃣  KEY STATISTICS:
    Pregnancies: min=0, max=17, mean=3

# Summary: What We Learned

## Dataset Overview
- **768 students** with **9 features** each
- **0.06 MB** memory footprint
- **0 missing values** (complete dataset)
- **0 duplicates (0.00%)** 

## Feature Types
- **1 Binary (Target Variable)**: Outcome 
- **6 Numeric Continuous**: Glucose, BloodPressure, SkinThickness, Insulin, BMI, DiabetetesPedigreeFunction
- **2 Numeric Discrete**: Pregnancies, Age

## Part 2: Data Dictionary (20 points)

Complete the following data dictionary. For each column, you must:
1. **Research** the clinical meaning
2. **Identify** the feature type (Continuous, Discrete, Categorical-Nominal, Categorical-Ordinal, Binary, Identifier)
3. **Document** the valid values/range you observe
4. **Note** any issues or questions